keras_cv/layers/object_detection3d/voxelization.py

```diff
@@ -0,0 +1,267 @@
+          point_voxel_feature: [B, N, dim] voxel feature (delta_{x,y,z}).
+          point_voxel_id: [B, N] voxel ID of each point. Invalid voxels have Id's
+            set to 0.
+          point_voxel_mask: [B, N] validpoint voxel boolean mask.
+        """
+        # [B, N, dim]
+        # convert from point coordinate to voxel index
+        point_voxelized = point_xyz / tf.constant(
+            self._voxel_size, dtype=point_xyz.dtype
+        )
+        point_voxelized_round = tf.math.round(point_voxelized)
+        point_voxel_xyz_float = tf.cast(point_voxelized_round, dtype=point_xyz.dtype)
+        # [B, N, dim]
+        # delta to the nearest voxel
+        point_voxel_feature = point_xyz - point_voxel_xyz_float * tf.constant(
+            self._voxel_size, dtype=point_xyz.dtype
+        )
+
+        # [B, N, dim]
+        point_voxel_xyz_int = tf.cast(point_voxel_xyz_float, dtype=tf.int32)
+        # [dim]
+        # get xmin, ymin, zmin
+        voxel_origin = self._spatial_size[::2]
+        voxel_origin = tf.constant(
+            [o / v for o, v in zip(voxel_origin, self._voxel_size)], dtype=tf.float32
+        )
+        voxel_origin = tf.math.round(voxel_origin)
+        voxel_origin = tf.cast(voxel_origin, dtype=tf.int32)
+
+        # [B, N, dim]
+        point_voxel_xyz = point_voxel_xyz_int - voxel_origin[tf.newaxis, tf.newaxis, :]
+
+        # [B, N]
+        # remove points outside of the voxel boundary
+        point_voxel_mask = tf.logical_and(
+            point_voxel_xyz >= 0,
+            point_voxel_xyz
+            < tf.constant(self._voxel_spatial_size, dtype=point_voxel_xyz.dtype),
+        )
+
+        # [B, N]
+        point_voxel_mask_int = tf.cast(point_voxel_mask, dtype=tf.int32)
+        # [B, N] for voxel_id, int constant for num_voxels, in the range of [0, B * num_voxels]
+        point_voxel_id = compute_point_voxel_id(
+            point_voxel_xyz, self._voxel_spatial_size
+        )
+        # [B, N]
+        point_voxel_id = point_voxel_id * point_voxel_mask_int
+
+        return point_voxel_feature, point_voxel_id, point_voxel_mask
+
+
+class Voxelization(tf.keras.layers.Layer):
+    """Dynamic voxelization and pool layer."""
+
+    def __init__(
+        self,
+        point_net: tf.keras.layers.Layer,
+        voxel_size: Sequence[float],
+        spatial_size: Sequence[float],
+        name: str = "voxelization",
+        **kwargs,
+    ):
+        super().__init__(name=name, **kwargs)
+        self._point_net = point_net
+        self._voxelization_layer = PointToVoxel(
+            voxel_size=voxel_size, spatial_size=spatial_size
+        )
+        self._voxel_size = voxel_size
+        self._spatial_size = spatial_size
+        self._voxel_spatial_size = compute_voxel_spatial_size(
+            spatial_size, self._voxel_size
+        )
+        self._voxel_spatial_size_volume = np.prod(self._voxel_spatial_size).item()
+
+    def call(
+        self,
+        point_xyz: tf.Tensor,
```

Comment:
Since only the first positional input to call is used to compute shape metadata in Keras, maybe the xyz+feature inputs should be passed in as a tuple instead of as different params?

Reply:
Yeah that'd be my proposal as well, I think this could potentially make `point_feature` more generic, it doesn't require a specific concatenation (much like our box concatenation over box / class / score today)